Trying to make an unsigned double?

I have a schema which looks like this:
```xml
<types>
   <composite name="MyCar">
     <!--<ref name="myfield" type="an_enum"/>-->
     <type name="speed" primitiveType="double"/>
  </composite>
</types>

<message name="foo">
  <group name="cars">
    <field name="myCar" id="1" type="MyCar"/>
  </group>
</message/>
```

And when trying to encode some data I get an err, which seems to be because we're trying to cast a double to an unsigned double. Encoding code:
```cpp
        auto d = m.cars();
        const auto group_size = 12;
        sbepp::fill_group_header(d, group_size);
        for (auto entry : d)
        {
          auto inner = entry.myCar();
          inner.myfield(MY_ENUM);
          inner.speed((double)argc); // <-- line of error
        }
```

Err:
```
root:1689:62: error: invalid use of incomplete type 'class std::__make_unsigned_selector<double, false, false>'
 1689 |     { typedef typename __make_unsigned_selector<_Tp>::__type type; };
      |                                                              ^~~~
root:1596:11: note: declaration of 'class std::__make_unsigned_selector<double, false, false>'
 1596 |     class __make_unsigned_selector;
      |           ^~~~~~~~~~~~~~~~~~~~~~~~
In file included from proj/build/sbeppc/my_sbe/schema/schema.hpp:7,
                 from proj/build/sbeppc/my_sbe/my_sbe.hpp:6,
                 from proj/flatbuf_sample_binary.cpp:29:
proj/sbepp/src/sbepp/sbepp.hpp: In instantiation of 'void sbepp::detail::set_primitive(Byte*, T) [with sbepp::endian E = sbepp::endian::little; T = double; Byte = char]':
proj/sbepp/src/sbepp/sbepp.hpp:605:21:   required from 'void sbepp::detail::set_value(View, std::size_t, T) [with sbepp::endian E = sbepp::endian::little; T = double; View = my_sbe::detail::types::composite_2<char>; std::size_t = long unsigned int]'
proj/build/sbeppc/my_sbe/types/PositionDelta.hpp:73:60:   required from 'void my_sbe::detail::types::composite_2<Byte>::amt(my_sbe::detail::types::type_4) const [with T = void; <template-parameter-2-2> = void; Byte = char]'
proj/flatbuf_sample_binary.cpp:495:38:   required from here
proj/sbepp/src/sbepp/sbepp.hpp:514:52: error: no matching function for call to 'to_unsigned(double&)'
  514 |         value = static_cast<T>(byteswap(to_unsigned(value)));
      |                                         ~~~~~~~~~~~^~~~~~~
proj/sbepp/src/sbepp/sbepp.hpp:361:48: note: candidate: 'template<class T> constexpr typename std::make_unsigned<_Tp>::type sbepp::detail::to_unsigned(T)'
  361 | constexpr typename std::make_unsigned<T>::type to_unsigned(T v) noexcept
      |                                                ^~~~~~~~~~~
proj/sbepp/src/sbepp/sbepp.hpp:361:48: note:   substitution of deduced template arguments resulted in errors seen above
```

Doubt this is related, but this is also happening. Might be a gcc9 thing.
```cpp
flatbuf_sample_binary.cpp:522:38: error: cannot convert 'sbepp::uint32_t' to 'uint32_t' {aka 'unsigned int'} in initialization
  522 |           uint32_t val = m.myUint32();
      |                          ~~~~~~~~~~^~
      |                                      |
      |                                      sbepp::uint32_t
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,9 @@
+# 1.0.1
+
+Fix floating-point types byte swapping for pre-C++20 compilers.
+
+---
+
 # 1.0.0
 
 Initial public release

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 3.11)
 
 project(sbepp
-    VERSION 1.0.0
+    VERSION 1.0.1
     LANGUAGES CXX
 )
 

--- a/doc/examples.md
+++ b/doc/examples.md
@@ -132,10 +132,10 @@ send(sbepp::addressof(m), msg_size);
 ```cpp
 std::array<char, 1024> buf{};
 auto header = sbepp::make_const_view<>(buf.data(), buf.size());
-if(header.templateId()
+if(*header.templateId()
     != sbepp::message_traits<market::schema::messages::msg>::id())
 {
-    std::cerr << "unknown message id: " << header.templateId() << '\n'; 
+    std::cerr << "unknown message id: " << *header.templateId() << '\n'; 
     return;
 }
 
@@ -190,7 +190,7 @@ for(const auto entry : g)
 std::array<char, 1024> buf{};
 auto m = sbepp::make_const_view<market::messages::msg>(buf.data(), buf.size());
 // there's nothing wrong in creation of a "wrong" message view
-if(sbepp::get_header(m).templateId()
+if(*sbepp::get_header(m).templateId()
     != sbepp::message_traits<market::schema::messages::msg>::id())
 {
     std::cerr << "not a `market::messages::msg`" << '\n'; 

--- a/sbepp/src/sbepp/sbepp.hpp
+++ b/sbepp/src/sbepp/sbepp.hpp
@@ -363,6 +363,52 @@ constexpr typename std::make_unsigned<T>::type to_unsigned(T v) noexcept
     return static_cast<typename std::make_unsigned<T>::type>(v);
 }
 
+inline std::uint64_t to_unsigned(double v) noexcept
+{
+    std::uint64_t res{};
+    std::memcpy(&res, &v, sizeof(res));
+    return res;
+}
+
+inline std::uint32_t to_unsigned(float v) noexcept
+{
+    std::uint32_t res{};
+    std::memcpy(&res, &v, sizeof(res));
+    return res;
+}
+
+template<typename To>
+struct from_unsigned
+{
+    constexpr To
+        operator()(typename std::make_unsigned<To>::type from) const noexcept
+    {
+        return static_cast<To>(from);
+    }
+};
+
+template<>
+struct from_unsigned<double>
+{
+    double operator()(std::uint64_t from) const noexcept
+    {
+        double res{};
+        std::memcpy(&res, &from, sizeof(res));
+        return res;
+    }
+};
+
+template<>
+struct from_unsigned<float>
+{
+    float operator()(std::uint32_t from) const noexcept
+    {
+        float res{};
+        std::memcpy(&res, &from, sizeof(res));
+        return res;
+    }
+};
+
 // intrinsics detection is taken from <boost/endian/detail/intrinsic.hpp>
 #ifndef __has_builtin          // Optional of course
 #    define __has_builtin(x) 0 // Compatibility with non-clang compilers
@@ -488,7 +534,7 @@ SBEPP_CPP20_CONSTEXPR T get_primitive(const Byte* ptr)
     }
     else
     {
-        return static_cast<T>(byteswap(to_unsigned(res)));
+        return from_unsigned<T>{}(byteswap(to_unsigned(res)));
     }
 #endif
 }
@@ -511,7 +557,7 @@ SBEPP_CPP20_CONSTEXPR void set_primitive(Byte* ptr, T value)
     // why explicit `std::memcpy` call is required
     if((E != endian::native) && (sizeof(T) != 1))
     {
-        value = static_cast<T>(byteswap(to_unsigned(value)));
+        value = from_unsigned<T>{}(byteswap(to_unsigned(value)));
     }
     std::memcpy(ptr, &value, sizeof(T));
 #endif

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -75,6 +75,7 @@ function(add_test_binary name_suffix cpp_version)
         ${src_dir}/size_bytes_checked.test.cpp
         ${src_dir}/borrowed_range.test.cpp
         ${src_dir}/stringification.test.cpp
+        ${src_dir}/float_fields.test.cpp
     )
 
     target_include_directories(${test_name}

--- a/test/schemas/big_endian_schema.xml
+++ b/test/schemas/big_endian_schema.xml
@@ -46,6 +46,14 @@
         <composite name="composite_3">
             <ref name="field" type="numbers_enum"/>
         </composite>
+
+        <composite name="composite_4">
+            <type name="field" primitiveType="float"/>
+        </composite>
+
+        <composite name="composite_5">
+            <type name="field" primitiveType="double"/>
+        </composite>
     </types>
 
     <sbe:message name="msg1" id="1">

--- a/test/schemas/test_schema.xml
+++ b/test/schemas/test_schema.xml
@@ -222,6 +222,11 @@
         <composite name="composite_14">
             <ref name="field" type="numbers_enum" offset="20"/>
         </composite>
+
+        <composite name="composite_19">
+            <type name="float_field" primitiveType="float"/>
+            <type name="double_field" primitiveType="double"/>
+        </composite>
     </types>
 
     <sbe:message name="Msg1" id="1">

--- a/test/src/sbepp/test/float_fields.test.cpp
+++ b/test/src/sbepp/test/float_fields.test.cpp
@@ -1,0 +1,42 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2023, Oleksandr Koval
+
+#ifdef USE_TOP_FILE
+#    include <test_schema/test_schema.hpp>
+#else
+#    include <test_schema/types/composite_19.hpp>
+#endif
+
+#include <gtest/gtest.h>
+
+#include <array>
+
+namespace
+{
+using byte_type = std::uint8_t;
+
+class FloatFieldsTest : public ::testing::Test
+{
+public:
+    std::array<byte_type, 128> buf{};
+    test_schema::types::composite_19<byte_type> c{buf.data(), buf.size()};
+};
+
+TEST_F(FloatFieldsTest, SetsFloatFieldCorrectly)
+{
+    static constexpr float valid_value = 1.2;
+
+    c.float_field(valid_value);
+
+    ASSERT_EQ(*c.float_field(), valid_value);
+}
+
+TEST_F(FloatFieldsTest, SetsDoubleFieldCorrectly)
+{
+    static constexpr double valid_value = 1.2;
+
+    c.double_field(valid_value);
+
+    ASSERT_EQ(*c.double_field(), valid_value);
+}
+} // namespace

--- a/test/src/sbepp/test/memory_access.test.cpp
+++ b/test/src/sbepp/test/memory_access.test.cpp
@@ -42,6 +42,8 @@
 #    include <big_endian_schema/types/composite_1.hpp>
 #    include <big_endian_schema/types/composite_2.hpp>
 #    include <big_endian_schema/types/composite_3.hpp>
+#    include <big_endian_schema/types/composite_4.hpp>
+#    include <big_endian_schema/types/composite_5.hpp>
 #    include <big_endian_schema/messages/msg1.hpp>
 #endif
 
@@ -129,6 +131,7 @@ class entry_wrapper
 public:
     using base_t =
         typename sbepp::group_traits<GroupTag>::template entry_type<byte_type>;
+
     constexpr entry_wrapper(byte_type* ptr, const std::size_t size)
         : base_t{ptr, size, sbepp::group_traits<GroupTag>::block_length()}
     {
@@ -233,9 +236,27 @@ template<typename T>
 using is_numeric_type = has_value<T>;
 
 template<typename T>
-enable_if_t<is_numeric_type<T>::value, T> get_value_to_set(T /*numeric*/)
+using is_integral_type = std::integral_constant<
+    bool,
+    has_value<T>::value && std::is_integral<typename T::value_type>::value>;
+
+template<typename T>
+using is_floating_point_type = std::integral_constant<
+    bool,
+    has_value<T>::value
+        && std::is_floating_point<typename T::value_type>::value>;
+
+template<typename T>
+enable_if_t<is_integral_type<T>::value, T> get_value_to_set(T /*integral*/)
 {
     return T{123};
+}
+
+template<typename T>
+enable_if_t<is_floating_point_type<T>::value, T>
+    get_value_to_set(T /*floating*/)
+{
+    return T{123.45};
 }
 
 template<typename T>
@@ -377,7 +398,9 @@ TYPED_TEST(
 using BigEndianFieldContainers = ::testing::Types<
     big_endian_schema::types::composite_1<byte_type>,
     big_endian_schema::types::composite_2<byte_type>,
-    big_endian_schema::types::composite_3<byte_type>>;
+    big_endian_schema::types::composite_3<byte_type>,
+    big_endian_schema::types::composite_4<byte_type>,
+    big_endian_schema::types::composite_5<byte_type>>;
 
 template<typename T>
 using BigEndianTest = FieldsContainer<T>;

--- a/test/src/sbepp/test/message.test.cpp
+++ b/test/src/sbepp/test/message.test.cpp
@@ -258,7 +258,8 @@ constexpr auto constexpr_test()
     message_t m;
     m = message_t{buf.data(), buf.size()};
     m = m;
-    m = std::move(m);
+    auto m2 = std::move(m);
+    (void)m2;
 
     sbepp::fill_message_header(m);
     sbepp::addressof(m);

